server/http/server.go

```diff
@@ -0,0 +1,74 @@
+package httpserver
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var log = logging.Logger("lassie/httpserver")
+
+type HttpServer struct {
+	cancel   context.CancelFunc
+	ctx      context.Context
+	listener net.Listener
+	server   *http.Server
+}
+
+func NewHttpServer(ctx context.Context, address string, port uint) (*HttpServer, error) {
+	addr := fmt.Sprintf("%s:%d", address, port)
+	listener, err := net.Listen("tcp", addr) // assigns a port if port is 0
+	if err != nil {
+		return nil, err
+	}
+
+	mux := http.NewServeMux()
+
+	ctx, cancel := context.WithCancel(ctx)
+	server := &http.Server{
+		Addr:        fmt.Sprintf(":%d", port),
+		BaseContext: func(listener net.Listener) context.Context { return ctx },
+		Handler:     mux,
```

Comment:
if you just want the server to be shutdown by CLI context, you can just add a parameter here:
```golang
BaseContext: func(listener net.Listener) context.Context {
	return ctx
},
```

Reply:
this can eliminate all the switching on whether the server shuts down or the ctx shutsdown first in daemon.go

Reply:
I don't quite understand what you mean by this